cli/fathom_web/extract.py

```diff
@@ -0,0 +1,144 @@
+import base64
+import mimetypes
+import pathlib
+import shutil
+import re
+
+from click import argument, command, option, Path
+
+
+BASE64_DATA_PATTERN = re.compile(r'(data:(?P<mime>[a-zA-Z0-9]+?/[a-zA-Z0-9\-.+]+?);(\s?charset=utf-8;)?base64,(?P<string>[a-zA-Z0-9+/=]+))')
+BASE_TAG_PATTERN = re.compile(r'<base.*?>')
+OLD_CSP = re.compile(r"default-src 'none'; img-src data:; media-src data:; style-src data: 'unsafe-inline'; font-src data:; frame-src data:")
+NEW_CSP = r"default-src 'none'; img-src 'self' data:; media-src 'self' data:; style-src 'self' data: 'unsafe-inline'; font-src 'self' data:; frame-src 'self' data:"
+# These are MIME types the `mimetypes` library doesn't recognize or gets wrong. Matches were found at:
+# https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Complete_list_of_MIME_types
+MIME_TYPE_TO_FILE_EXTENSION = {
+    'application/font-woff': '.woff',
+    'application/font-woff2': '.woff2',
+    'application/vnd.ms-fontobject': '.eot',
+    'font/opentype': '.otf',
+    'font/ttf': '.ttf',
+    'font/woff2': '.woff2',
+    'image/jpeg': '.jpg',
+    'image/webp': '.webp',
+}
+
+
+@command()
+@option('--preserve-originals/--no-preserve-originals',
+        default=True,
+        help='Save original html files in a newly created `originals` directory in IN_DIRECTORY (default: True)')
+@argument('in_directory', type=Path(exists=True, file_okay=False))
+def main(in_directory, preserve_originals):
+    """
+    Extracts resources from the html pages in IN_DIRECTORY and stores them in a separate directory for Git-LFS storage.
+    """
+    if preserve_originals:
+        originals_dir = pathlib.Path(in_directory) / 'originals'
+        originals_dir.mkdir(parents=True, exist_ok=True)
+
+    for file in pathlib.Path(in_directory).iterdir():
+        if file.suffix != '.html':
+            print(f'Skipping {file.name}; not an html file')
+            continue
+        if preserve_originals:
+            shutil.copyfile(file, originals_dir / file.name)
+        extract_base64_data_from_html_page(file)
+
+
+def extract_base64_data_from_html_page(file: pathlib.Path):
+    """
+    Extract all base64 data from the given html page and store the data in separate files.
+
+    We do this by building a new html string using the non-base64 data pieces from the original file and the
+    filenames we will generate for each of the base64 data strings.
+
+    Base64 data is found with regex matching. Each data string is decoded and saved as a separate file.
+    """
+    with file.open(encoding='utf-8') as fp:
+        html = fp.read()
+
+    # Make the subresources directory
+    subresources_directory = file.parent / 'resources' / f'{file.stem}_resources'
```

Comment:
Is there a purpose for the "_resources" suffix? If not, I'd do without to save a little reading, typing, and space. It's already within a folder called "resources".

Reply:
The drag is that, if we stick the `resources` dir next to the samples being processed, we can't really move the samples afterward without breaking the references. I envision us running fathom-extract as a first step, as soon as a corpus is collected, because otherwise we can't check it in anywhere. Then, when we use fathom-pick to grab out some training samples, their resources won't come with them. The obvious solutions I see are…

1. Put a single resources dir somewhere, like at the root of the samples dir. Of course, this means we can never move samples any shallower or deeper in the folder hierarchy.
2. Do (1), but also stick a symlink to the root-level resources dir in each sample dir (training, validation, etc.). Git will store symlinks cluefully, but I'm not sure how they check out on Windows.
3. Do (1), but teach fathom-serve about a special `resources` dir (still living at the root level), and have it expose said dir everywhere, so validation/resources will work, and training/resources will work, but they'll point to the same dir on disk. This will complicate things for the user, because they'll either need to pass fathom-serve the location of the resources dir, or fathom-serve will have to track it down itself according to some conventions, which will surprise someone at some point.

My favorite is (2), at least if it works on Windows. I've used a similar trick for letting PHP apps find their shared includes, and it worked well.

Any other alternatives?

Reply:
Hmm, symlinks on Windows do not look promising: https://github.com/git-for-windows/git/wiki/Symbolic-Links. (I know we experimented with it a few weeks back, but I forgot how it worked out.) But wait, maybe they just started working! https://www.quora.com/How-does-Git-handle-symbolic-links-on-Windows

Reply:
The quora page you linked makes it seem there are still significant hoops a Windows user would have to jump through for symlinks to work. What do you think about storing resources in separate directories next to the pages? We could change `fathom-pick` to move resource directories. It will probably make setting up a repo and git-LFS a little more difficult, but I imagine we'll already need to supply a guide for users for that.

Reply:
We decided to make changes to fathom-pick to also move resources. This means Windows people don't need to mess with their machines and we still can just double-click samples to open them if you futz with your FF CORS settings.